### PR TITLE
Ftr:  zk/consul Metadata

### PR DESCRIPTION
**Pull request by gaoxinge**
**What this PR does**:

- add consul metadata report
- add zookeeper metadata report

**Which issue(s) this PR fixes**:

Fixes https://github.com/apache/dubbo-go/issues/445.

**Special notes for your reviewer**:

- TODO
    - [x] add consul metadata report unit test
    - [x] add zookeeper metadata report unit test

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -20,15 +20,13 @@ classes
 
 # go mod, go test
 vendor/
-coverage.txt
-
 logs/
 .vscode/
-coverage.txt
```
Comment (zouyx) [inline .gitignore:27]: why delete this line？

File: metadata/report/consul/report.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+// GetExportedURLs gets the urls.
+func (m *consulMetadataReport) GetExportedURLs(metadataIdentifier *identifier.ServiceMetadataIdentifier) []string {
+	k := metadataIdentifier.GetIdentifierKey()
+	kv, _, err := m.client.KV().Get(k, nil)
+	if err != nil {
+		panic(err)
+	}
+
+	if kv == nil {
+		return []string{}
```
Comment (zouyx) [inline metadata/report/consul/report.go:82]: Suggest use constant `[]string` for return

File: metadata/report/consul/report.go [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+// GetSubscribedURLs gets the urls.
+func (m *consulMetadataReport) GetSubscribedURLs(subscriberMetadataIdentifier *identifier.SubscriberMetadataIdentifier) []string {
+	k := subscriberMetadataIdentifier.GetIdentifierKey()
+	kv, _, err := m.client.KV().Get(k, nil)
+	if err != nil {
+		panic(err)
+	}
+
+	if kv == nil {
+		return []string{}
```
Comment (zouyx) [inline metadata/report/consul/report.go:103]: As above

File: metadata/report/nacos/report.go [hunk 4]
```diff
@@ -39,19 +38,19 @@ import (
...
 }
 
-// nacosMetadataReport is the implementation of MetadataReport based Nacos
+// nacosMetadataReport is the implementation
+// of MetadataReport based on nacos.
 type nacosMetadataReport struct {
 	client config_client.IConfigClient
 }
 
-// StoreProviderMetadata will store the metadata
```
Comment (zouyx) [inline metadata/report/nacos/report.go:53]: why change all comments of this file?

File: .gitignore (hunk 1 above)
Comment (gaoxinge) [reply .gitignore:27]: These `coverage.txt` are duplicated with https://github.com/apache/dubbo-go/blob/ecf7fa15689c0969384602a2506dc17fc87d4fe1/.gitignore#L14.

File: metadata/report/consul/report.go (hunk 2 above)
Comment (gaoxinge) [reply metadata/report/consul/report.go]: I replace it with global variable `emptyStrSlice`.

File: metadata/report/consul/report.go (hunk 3 above)
Comment (gaoxinge) [reply metadata/report/consul/report.go]: I replace it with global variable `emptyStrSlice`.

File: metadata/report/nacos/report.go (hunk 4 above)
Comment (gaoxinge) [reply metadata/report/nacos/report.go:53]: Unify the comment for the same functions in different metadata report.

File: metadata/report/consul/report.go [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+
+var (
+	emptyStrSlice = make([]string, 0)
+)
+
+func init() {
+	mf := &consulMetadataReportFactory{}
+	extension.SetMetadataReportFactory("consul", func() factory.MetadataReportFactory {
+		return mf
+	})
```
Comment (hxmhlt) [inline metadata/report/consul/report.go:40]: combine two lines.

File: metadata/report/consul/report.go [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+
+// consulMetadataReport is the implementation of
+// MetadataReport based on consul.
+type consulMetadataReport struct {
+	client *consul.Client
+}
+
+// StoreProviderMetadata stores the metadata.
+func (m *consulMetadataReport) StoreProviderMetadata(providerIdentifier *identifier.MetadataIdentifier, serviceDefinitions string) error {
+	kv := &consul.KVPair{Key: providerIdentifier.GetIdentifierKey(), Value: []byte(serviceDefinitions)}
```
Comment (hxmhlt) [inline metadata/report/consul/report.go:51]: go fmt not work ?

File: metadata/report/consul/report.go [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+	}
+	return []string{string(kv.Value)}
+}
+
+// GetServiceDefinition gets the service definition.
+func (m *consulMetadataReport) GetServiceDefinition(metadataIdentifier *identifier.MetadataIdentifier) string {
+	k := metadataIdentifier.GetIdentifierKey()
+	kv, _, err := m.client.KV().Get(k, nil)
+	if err != nil {
+		panic(err)
```
Comment (hxmhlt) [inline metadata/report/consul/report.go:117]: don't panic

File: metadata/report/delegate/delegate_report.go [hunk 8]
```diff
@@ -241,11 +242,20 @@ func (mr *MetadataReport) GetExportedURLs(identifier *identifier.ServiceMetadata
...
+	urlStrList := make([]string, 0, len(urls))
+	for _, url := range urls {
+		urlStrList = append(urlStrList, url.String())
+	}
+	bytes, err := json.Marshal(urlStrList)
+	if err != nil {
+		return perrors.WithMessage(err, "Could not convert the array to json")
+	}
+
 	report := instance.GetMetadataReportInstance()
```
Comment (hxmhlt) [inline metadata/report/delegate/delegate_report.go:254]: This common logic may be confirmed by @flycash

Comment (hxmhlt) [reply metadata/report/delegate/delegate_report.go:254]: May create influence to nacos's implement.

File: metadata/report/nacos/report.go [hunk 9]
```diff
@@ -39,19 +38,19 @@ import (
 )
 
 func init() {
-	ftry := &nacosMetadataReportFactory{}
+	mf := &nacosMetadataReportFactory{}
 	extension.SetMetadataReportFactory("nacos", func() factory.MetadataReportFactory {
-		return ftry
+		return mf
 	})
```
Comment (hxmhlt) [inline metadata/report/nacos/report.go:44]: Combine

File: metadata/report/zookeeper/report.go [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+
+var (
+	emptyStrSlice = make([]string, 0)
+)
+
+func init() {
+	mf := &zookeeperMetadataReportFactory{}
+	extension.SetMetadataReportFactory("zookeeper", func() factory.MetadataReportFactory {
+		return mf
+	})
```
Comment (hxmhlt) [inline metadata/report/zookeeper/report.go:43]: combine

File: remoting/consul/agent.go [hunk 11]
```diff
@@ -0,0 +1,70 @@
+/*
```
Comment (hxmhlt) [inline remoting/consul/agent.go:1]: What the consultAgent used for ?

Comment (hxmhlt) [reply remoting/consul/agent.go:1]: If used for unit test, it's better to move to xxx_test file.

File: metadata/report/consul/report.go (hunk 5 above)
Comment (gaoxinge) [reply metadata/report/consul/report.go:39]: I have combined theres lines.

File: metadata/report/nacos/report.go (hunk 9 above)
Comment (gaoxinge) [reply metadata/report/nacos/report.go:43]: I have combined theres lines.

File: metadata/report/zookeeper/report.go (hunk 10 above)
Comment (gaoxinge) [reply metadata/report/zookeeper/report.go:42]: I have combined theres lines.

File: remoting/consul/agent.go (hunk 11 above)
Comment (gaoxinge) [reply remoting/consul/agent.go]: This `consulAgent` is used for test, and it is used for test and imported by test. I rename `agent.go` to `test_agent.go`.

File: metadata/report/consul/report.go (hunk 7 above)
Comment (gaoxinge) [reply metadata/report/consul/report.go]: I remove this panic and return an error in this function.

File: metadata/report/consul/report.go (hunk 6 above)
Comment (gaoxinge) [reply metadata/report/consul/report.go:50]: It is ok with `go fmt`.

File: metadata/report/delegate/delegate_report.go (hunk 8 above)
Comment (gaoxinge) [reply metadata/report/delegate/delegate_report.go:254]: I think this is a common code for different metadata report, and one can find the [similar code](https://github.com/apache/dubbo/blob/master/dubbo-metadata/dubbo-metadata-api/src/main/java/org/apache/dubbo/metadata/report/support/AbstractMetadataReport.java#L320-L327) in java version.

File: metadata/report/consul/report.go [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+	k := metadataIdentifier.GetIdentifierKey()
+	_, err := m.client.KV().Delete(k, nil)
+	return err
+}
+
+// GetExportedURLs gets the urls.
+func (m *consulMetadataReport) GetExportedURLs(metadataIdentifier *identifier.ServiceMetadataIdentifier) ([]string, error) {
+	k := metadataIdentifier.GetIdentifierKey()
+	kv, _, err := m.client.KV().Get(k, nil)
+	if err != nil {
```
Comment (AlexStocks) [inline metadata/report/consul/report.go:80]: if err != nil ||  kv == nil {
		return emptyStrSlice, err
	}

File: metadata/report/consul/report.go [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+func (m *consulMetadataReport) SaveSubscribedData(subscriberMetadataIdentifier *identifier.SubscriberMetadataIdentifier, urls string) error {
+	kv := &consul.KVPair{Key: subscriberMetadataIdentifier.GetIdentifierKey(), Value: []byte(urls)}
+	_, err := m.client.KV().Put(kv, nil)
+	return err
+}
+
+// GetSubscribedURLs gets the urls.
+func (m *consulMetadataReport) GetSubscribedURLs(subscriberMetadataIdentifier *identifier.SubscriberMetadataIdentifier) ([]string, error) {
+	k := subscriberMetadataIdentifier.GetIdentifierKey()
+	kv, _, err := m.client.KV().Get(k, nil)
```
Comment (AlexStocks) [inline metadata/report/consul/report.go:100]: if err != nil ||  kv == nil {
	return emptyStrSlice, err
}

File: metadata/report/consul/report.go [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+		return emptyStrSlice, nil
+	}
+	return []string{string(kv.Value)}, nil
+}
+
+// GetServiceDefinition gets the service definition.
+func (m *consulMetadataReport) GetServiceDefinition(metadataIdentifier *identifier.MetadataIdentifier) (string, error) {
+	k := metadataIdentifier.GetIdentifierKey()
+	kv, _, err := m.client.KV().Get(k, nil)
+	if err != nil {
```
Comment (AlexStocks) [inline metadata/report/consul/report.go:115]: if err != nil ||  kv == nil {
	return emptyStrSlice, err
}

File: metadata/report/consul/report.go [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+	if kv == nil {
+		return "", nil
+	}
+	return string(kv.Value), nil
+}
+
+type consulMetadataReportFactory struct {
+}
+
+func (mf *consulMetadataReportFactory) CreateMetadataReport(url *common.URL) report.MetadataReport {
```
Comment (AlexStocks) [inline metadata/report/consul/report.go:128]: If u do not wanna add comment for this func, pls add "// nolint".

File: metadata/report/delegate/delegate_report.go (hunk 8 above)
Comment (flycash) [reply metadata/report/delegate/delegate_report.go:254]: Let's keep those codes, I will test it again

File: metadata/report/nacos/report.go (hunk 9 above)
Comment (flycash) [reply metadata/report/nacos/report.go:43]: This is different.
If the ftry is outside the function, it means that we always return the same instance.
But with this change, it means that we create an instance of factory.
Actually, the MetadataReportFactory should be singleton.
So you can not combine these codes.

File: metadata/report/zookeeper/report.go [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+func (m *zookeeperMetadataReport) RemoveServiceMetadata(metadataIdentifier *identifier.ServiceMetadataIdentifier) error {
+	k := m.rootDir + metadataIdentifier.GetFilePathKey()
+	return m.client.Delete(k)
+}
+
+// GetExportedURLs gets the urls.
+func (m *zookeeperMetadataReport) GetExportedURLs(metadataIdentifier *identifier.ServiceMetadataIdentifier) ([]string, error) {
+	k := m.rootDir + metadataIdentifier.GetFilePathKey()
+	v, _, err := m.client.GetContent(k)
+	if err != nil {
```
Comment (flycash) [inline metadata/report/zookeeper/report.go:80]: if err != nil || len(v2) == 0 
  return emptyStrSlice, nil

File: metadata/report/zookeeper/report.go [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+func (m *zookeeperMetadataReport) SaveSubscribedData(subscriberMetadataIdentifier *identifier.SubscriberMetadataIdentifier, urls string) error {
+	k := m.rootDir + subscriberMetadataIdentifier.GetFilePathKey()
+	return m.client.CreateWithValue(k, []byte(urls))
+}
+
+// GetSubscribedURLs gets the urls.
+func (m *zookeeperMetadataReport) GetSubscribedURLs(subscriberMetadataIdentifier *identifier.SubscriberMetadataIdentifier) ([]string, error) {
+	k := m.rootDir + subscriberMetadataIdentifier.GetFilePathKey()
+	v, _, err := m.client.GetContent(k)
+	if err != nil {
```
Comment (flycash) [inline metadata/report/zookeeper/report.go:100]: see above comment

File: metadata/report/consul/report.go (hunk 5 above)
Comment (gaoxinge) [reply metadata/report/consul/report.go:40]: I keep this code for singleton.

File: metadata/report/nacos/report.go (hunk 9 above)
Comment (gaoxinge) [reply metadata/report/nacos/report.go:44]: @hxmhlt @flycash I keep this code for singleton.

File: metadata/report/zookeeper/report.go (hunk 10 above)
Comment (gaoxinge) [reply metadata/report/zookeeper/report.go:43]: I keep this code for singleton.

File: metadata/report/consul/report.go (hunk 12 above)
Comment (gaoxinge) [reply metadata/report/consul/report.go]: I follow your advice and keep code more simpler.

File: metadata/report/consul/report.go (hunk 13 above)
Comment (gaoxinge) [reply metadata/report/consul/report.go:97]: I follow your advice and keep code more simpler.

File: metadata/report/consul/report.go (hunk 14 above)
Comment (gaoxinge) [reply metadata/report/consul/report.go]: I follow your advice and keep code more simpler.

File: metadata/report/consul/report.go (hunk 15 above)
Comment (gaoxinge) [reply metadata/report/consul/report.go:118]: I add `// nolint` in every `CreateMetadataReport`.

File: metadata/report/zookeeper/report.go (hunk 16 above)
Comment (gaoxinge) [reply metadata/report/zookeeper/report.go]: I follow your advice and keep code more simpler.

File: metadata/report/zookeeper/report.go (hunk 17 above)
Comment (gaoxinge) [reply metadata/report/zookeeper/report.go]: I follow your advice and keep code more simpler.

File: metadata/report/consul/report_test.go [hunk 18]
```diff
@@ -0,0 +1,166 @@
...
+)
+
+func newProviderRegistryUrl(host string, port int) *common.URL {
+	url1 := common.NewURLWithOptions(
+		common.WithIp(host),
+		common.WithPort(strconv.Itoa(port)),
+		common.WithParams(url.Values{}),
+		common.WithParamsValue(constant.ROLE_KEY, strconv.Itoa(common.PROVIDER)),
+	)
+	return url1
```
Comment (zouyx) [inline metadata/report/consul/report_test.go:47]:
```suggestion
	return common.NewURLWithOptions(
		common.WithIp(host),
		common.WithPort(strconv.Itoa(port)),
		common.WithParams(url.Values{}),
		common.WithParamsValue(constant.ROLE_KEY, strconv.Itoa(common.PROVIDER)),
	)
```

what about this?

File: metadata/report/report.go [hunk 19]
```diff
@@ -28,8 +28,8 @@ type MetadataReport interface {
...
 	SaveServiceMetadata(*identifier.ServiceMetadataIdentifier, common.URL) error
 	RemoveServiceMetadata(*identifier.ServiceMetadataIdentifier) error
-	GetExportedURLs(*identifier.ServiceMetadataIdentifier) []string
-	SaveSubscribedData(*identifier.SubscriberMetadataIdentifier, []common.URL) error
-	GetSubscribedURLs(*identifier.SubscriberMetadataIdentifier) []string
-	GetServiceDefinition(*identifier.MetadataIdentifier) string
+	GetExportedURLs(*identifier.ServiceMetadataIdentifier) ([]string, error)
+	SaveSubscribedData(*identifier.SubscriberMetadataIdentifier, string) error
+	GetSubscribedURLs(*identifier.SubscriberMetadataIdentifier) ([]string, error)
+	GetServiceDefinition(*identifier.MetadataIdentifier) (string, error)
```
Comment (zouyx) [inline metadata/report/report.go:34]: add some comments for this file pls.

File: metadata/report/zookeeper/report_test.go [hunk 20]
```diff
@@ -0,0 +1,167 @@
...
+)
+
+func newProviderRegistryUrl(host string, port int) *common.URL {
+	url1 := common.NewURLWithOptions(
+		common.WithIp(host),
+		common.WithPort(strconv.Itoa(port)),
+		common.WithParams(url.Values{}),
+		common.WithParamsValue(constant.ROLE_KEY, strconv.Itoa(common.PROVIDER)),
+	)
+	return url1
```
Comment (zouyx) [inline metadata/report/zookeeper/report_test.go:47]:
```suggestion
	return common.NewURLWithOptions(
		common.WithIp(host),
		common.WithPort(strconv.Itoa(port)),
		common.WithParams(url.Values{}),
		common.WithParamsValue(constant.ROLE_KEY, strconv.Itoa(common.PROVIDER)),
	)
```

File: metadata/report/report.go (hunk 19 above)
Comment (gaoxinge) [reply metadata/report/report.go]: I shamelessly copy the [comments](https://github.com/apache/dubbo-go/blob/feature/dubbo-2.7.5/metadata/report/nacos/report.go) in original nacos metadata report.

File: metadata/report/zookeeper/report_test.go (hunk 20 above)
Comment (gaoxinge) [reply metadata/report/zookeeper/report_test.go]: I simplify some returns in my code.

File: metadata/report/consul/report_test.go (hunk 18 above)
Comment (gaoxinge) [reply metadata/report/consul/report_test.go]: I simplify some returns in my code.

**Pull request merged**